### PR TITLE
Selectively disable compression based on the incoming request. Fixes #10377

### LINKED ISSUE #10377: Meteor Auto-compress interfering with Brotli enabled nginx front-end proxy
Hello @benjamn,

the Auto-compress feature of Meteor currently uses gzip compression and is enabled by default.
https://github.com/meteor/meteor/blob/release-1.8.1/packages/webapp/webapp_server.js#L822

When using a front-end proxy, such as nginx, it is not possible to use a different Content-Encoding such as Brotli, because the files are already gzip encoded.

I would like to discuss the possibility to introduce a new environment variable, that allows disabling the Auto-compress feature:
`METEOR_DISABLE_AUTO_COMPRESS=1`

The implementation is in PR: #10378

**1) Running Meteor with --production flag:**
gzip is used for Content-Encoding.
01-meteor-gzip

**2) Running Meteor with --production flag behind nginx proxy with Brotli enabled:**
Mixed Content-Encoding with gzip and br. The gzip encoded files cannot be encoded to Brotli.
02-meteor-nginx-mixed

**3) Running patched Meteor with --production flag behind nginx proxy with Brotli enabled:**
Only Brotli is used for Content-Encoding. nginx takes care of encoding the uncompressed files to Brotli.
03-meteor-nginx-brotli

**Related work on Brotli support**
Currently Node.js is working actively on adding Brotli compression to core:
https://github.com/nodejs/node/pull/20458
https://github.com/nodejs/node/issues/18964

There were some discussions in https://github.com/expressjs/compression to support Brotli compression, that were not further pursued.

Brotli browser support: https://caniuse.com/#feat=brotli

**Reasoning for environment variable**
Adding an environment variable to disable Auto-compress will allow to use Brotli compression through a front-end proxy and provide better compression ratio than gzip.

**Some tests with my PR (#10378)**

| Project  | gzip  | brotli | br (lvl 11) | decrease | decrease (br lvl 11) |
| -------- | ----- | ------ | ----------- | -------- | -------------------- |
| minimal  | 16.2  | 15.2   | 14          |  -6.17%  | -13.58%              |
| bare     | 65.3  | 61     | 55.8        |  -6.58%  | -14.55%              |
| full     | 130   | 122    | 112         |  -6.15%  | -13.85%              |
| react    | 106   | 99.1   | 90.6        |  -6.51%  | -14.53%              |
| angular  | 255   | 233    | 210         |  -8.63%  | -17.64%              |
| apollo   | 385   | 351    | 315         |  -8.83%  | -18.18%              |

Note:
- size in KB
- Brotli default is level 6 (ngx_brotli levels: 0-11)
- Tested on modern browser bundle with --production flag
- angular: https://github.com/Urigo/angular-meteor/tree/master/examples/MeteorCLI/all-in-one
- apollo: https://github.com/Urigo/angular-meteor/tree/master/examples/apollo

Test environment:
Ubuntu 18.04
nginx-1.15.7
brotli: 1.0.3-1ubuntu1.2
ngx_brotli: https://github.com/eustas/ngx_brotli

Thanks, Georgy

### PR DESCRIPTION
**Pull request by georgyberdyshev**
Hello @benjamn,

this a PR to fix https://github.com/meteor/meteor/issues/10377

Currently Auto-compress forces gzip compression, that cannot be disabled and therefore interferes with the possibility to enable a different Content-Encoding on a frontend-proxy.

Adding an environment variable to disable Auto-compress will allow to use Brotli compression through a front-end proxy, such as nginx, and provide better compression ratio than gzip.

`METEOR_DISABLE_AUTO_COMPRESS=1`

For more details, please see:
https://github.com/meteor/meteor/issues/10377

Thanks, Georgy

### CODE REVIEW

**Comment by benjamn** [comment]
@georgyberdyshev We're open to this solution (since it will solve your use case), but I was just wondering if your nginx proxy is stripping/disabling the [`Accept-Encoding`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Encoding) header before forwarding the request to the Meteor server? Ideally the Meteor server (i.e. the `compress()` middleware) would honor (the absence of) that header, rather than blindly compressing everything. Alternatively/additionally, we could [configure the compression middleware with a filter](https://www.npmjs.com/package/compression#filter-1) that would selectively disable compression based on the incoming request.

**Comment by georgyberdyshev** [comment]
Hello @benjamn,

thank you for your reply. Meteor behind the nginx proxy correctly receives the Accept-Encoding request HTTP header:

nginx log:
`10.0.2.2 - - [21/Dec/2018:17:40:51 +0000] "GET /headers HTTP/2.0" 200 595 "-" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.28 Safari/537.36" "-" "gzip, deflate, br"`

Meteor:
```
{ 'x-forwarded-host': 'localhost',
  'x-forwarded-proto': 'http',
  'x-forwarded-port': '80',
  'accept-language': 'en-US,en;q=0.9',
  'accept-encoding': 'gzip, deflate, br',
  accept: 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8',
  'user-agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.28 Safari/537.36',
  'upgrade-insecure-requests': '1',
  'cache-control': 'max-age=0',
  'x-no-compression': 'x-no-compression',
  connection: 'close',
  host: 'localhost',
  'x-forwarded-for': '10.0.2.2,127.0.0.1' }
```

I have updated the PR to use the x-no-compression HTTP request header, as this will provide more granularity and also makes and additional environment variable for Meteor obsolete.

Thanks, Georgy

File: packages/webapp/webapp_server.js [hunk 1]
```diff
@@ -64,6 +64,16 @@ var sha1 = function (contents) {
   return hash.digest('hex');
 };
 
+var shouldCompress = function (req, res) {
```
Comment (benjamn) [inline packages/webapp/webapp_server.js]:
```suggestion
 function shouldCompress(req, res) {
```

**Comment by georgyberdyshev** [comment]
Thanks @benjamn !

**Comment by amsinha2** [comment]
Im trying to enable brotli with meteor 2.7.3 Is this environment variable still relevant here?

**Pull request merged**